src/Resource/ParameterBag.js

```diff
@@ -56,81 +49,99 @@ ParameterBag.prototype = {
    * @returns {HyralFilter[]}
    */
   get filters() {
-    return this.parameters.filters;
+    return this._state.parameters.filters;
   },
 
   /**
    * @param {HyralFilter} filter
    */
   addFilter(filter) {
-    this.parameters.filters.push(filter);
+    this._state.parameters.filters.push(filter);
     this.updateStateId();
   },
 
   /**
    * @param {HyralFilter[]} filters
    */
   setFilters(filters) {
-    this.parameters.filters = filters;
+    this._state.parameters.filters = filters;
     this.updateStateId();
   },
 
   /**
    * @returns {HyralPaging}
    */
   get paging() {
-    return this.parameters.paging;
+    return this._state.parameters.paging;
   },
 
   /**
    * @param {HyralPaging} paging
    */
   setPaging(paging) {
-    this.parameters.paging = paging;
+    this._state.parameters.paging = paging;
     this.updateStateId();
   },
 
   /**
    * @returns {HyralSorting[]}
    */
   get sorting() {
-    return this.parameters.sorting;
+    return this._state.parameters.sorting;
   },
 
   /**
    * @param {HyralSorting[]} sorting
    */
   setSorting(sorting) {
-    this.parameters.sorting = sorting;
+    this._state.parameters.sorting = sorting;
     this.updateStateId();
   },
 
   /**
    * @returns {Object}
    */
   get params() {
-    return this.parameters.params;
+    return this._state.parameters.params;
   },
 
   /**
    * @param {string} key
    * @param value
    */
   addParam(key, value) {
-    this.parameters.params[key] = value;
+    this._state.parameters.params[key] = value;
     this.updateStateId();
   },
 
   /**
    * @param {Object} params
    */
   setParams(params) {
-    this.parameters.params = params;
+    this._state.parameters.params = params;
     this.updateStateId();
   },
 
   get stateId() {
-    return this.metadata.stateId;
+    return this._state.metadata.stateId;
+  },
+
+  updateStateId() {
+    this._state.metadata.stateId += 1;
+  },
+
+  /**
+   * @returns {object}
+   */
+  get state() {
+    return cloneDeep(this._state);
+  },
+
+  /**
+   * @param newState {object}
+   */
+  set newState(newState) {
```

Comment:
Waarom is het hier newState en niet gewoon state?

Reply:
Ik had inderdaad in eerste instantie state staan. 
Mijn redenering was dat je op deze manier heel bewust de state aanpast.